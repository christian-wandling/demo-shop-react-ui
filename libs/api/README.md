# api

A comprehensive library for consuming the demo-shop api. Supports accessing a product catalog, user management,
shopping cart operations, and order processing.
