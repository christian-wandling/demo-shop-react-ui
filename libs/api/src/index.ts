export * from './lib/openapi';
export * from './lib/ApiContext';
export * from './lib/services/apiService';
export * from './lib/ApiProvider';
export * from './lib/services/apiService';
export * from './lib/middleware/authMiddleware';
