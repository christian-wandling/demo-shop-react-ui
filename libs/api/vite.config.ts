import { defineConfig } from 'vite';
import react from '@vitejs/plugin-react';

export default defineConfig(() => ({
  root: __dirname,
  cacheDir: '../../node_modules/.vite/libs/api',
  plugins: [react()],
  // Uncomment this if you are using workers.
  // worker: {
  //  plugins: [ nxViteTsPaths() ],
  // },
  test: {
    watch: false,
    globals: true,
    environment: 'jsdom',
    reporters: ['default'],
    passWithNoTests: true,
    coverage: {
      reportsDirectory: './test-output/vitest/coverage',
      enabled: false,
    },
  },
}));
