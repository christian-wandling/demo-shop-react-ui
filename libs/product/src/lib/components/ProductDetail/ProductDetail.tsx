import { formatCurrency } from '@demo-shop-react-ui/shared';
import { useShoppingCartStore } from '@demo-shop-react-ui/shopping';
import { useProductStore } from '../../+state/useProductStore';
import { useEffect } from 'react';
import { useParams } from 'react-router';

export const ProductDetail = () => {
  const { id } = useParams();
  const getProductById = useProductStore(state => state.getProductById);
  const fetchProductById = useProductStore(state => state.fetchProductById);
  const hasActiveSession = useShoppingCartStore(state => state.hasActiveSession);
  const add = useShoppingCartStore(state => state.add);

  useEffect(() => {
    if (id) {
      fetchProductById(+id);
    }
  }, [id, fetchProductById]);

  if (!id) {
    return <div>Product ID is missing</div>;
  }

  const product = getProductById(+id);

  if (!product) {
    return <div>Loading...</div>;
  }

  return (
    <div className="product-detail pt-6">
      <div className="mx-auto mt-6 max-w-2xl sm:px-6 lg:grid lg:max-w-7xl lg:grid-cols-3 lg:gap-x-8 lg:px-8">
        <div className="aspect-h-1 aspect-w-1 overflow-hidden sm:rounded-lg lg:block">
          <img
            src={product.thumbnail.uri ?? ''}
            alt={product.name}
            onError={e => {
              e.currentTarget.src = 'images/placeholder-image.jpg';
            }}
            className="h-full w-full object-cover object-center group-hover:opacity-75"
          />
        </div>
      </div>

      <div className="mx-auto max-w-2xl px-4 pb-16 pt-10 sm:px-6 lg:grid lg:max-w-7xl lg:grid-cols-3 lg:grid-rows-[auto,auto,1fr] lg:gap-x-8 lg:px-8 lg:pb-24 lg:pt-16">
        <div className="lg:col-span-2 lg:border-r lg:border-gray-200 lg:pr-8">
          <h1 className="text-2xl font-bold tracking-tight text-gray-900 sm:text-3xl">{product.name}</h1>
        </div>

        <div className="mt-4 lg:row-span-3 lg:mt-0">
          <h2 className="sr-only">Product information</h2>
          <p className="text-3xl tracking-tight text-gray-900">{formatCurrency(product.price)}</p>

          <button
            onClick={_ => add({ productId: product.id })}
            disabled={!hasActiveSession()}
            className={`btn-add mt-10 flex w-full items-center justify-center rounded-md border border-transparent px-8 py-3 text-base font-medium
  ${
    hasActiveSession()
      ? 'bg-indigo-600 text-white hover:bg-indigo-700 focus:outline-none focus:ring-2 focus:ring-indigo-500 focus:ring-offset-2'
      : 'bg-gray-100 text-gray-800 pointer-events-none'
  }`}>
            Add to cart
          </button>
        </div>

        <div className="py-10 lg:col-span-2 lg:col-start-1 lg:border-r lg:border-gray-200 lg:pb-16 lg:pr-8 lg:pt-6">
          <div className="product-description text-sm text-gray-600">{product.description}</div>
        </div>
      </div>
    </div>
  );
};
