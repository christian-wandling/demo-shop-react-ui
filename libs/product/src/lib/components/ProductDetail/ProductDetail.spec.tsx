import { beforeEach, describe, expect, it, vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import { ProductDetail } from './ProductDetail';
import { ProductResponse } from '@demo-shop-react-ui/api';

vi.mock('react-router', () => ({
  useParams: vi.fn().mockReturnValue({ id: '1' }),
}));

const fetchProductByIdSpy = vi.fn();
const getProductByIdSpy = vi.fn();

vi.mock('../../+state/useProductStore', () => ({
  useProductStore: vi.fn().mockImplementation(selector => {
    const store = {
      fetchProductById: fetchProductByIdSpy,
      getProductById: getProductByIdSpy,
    };
    return selector(store);
  }),
}));

vi.mock('@demo-shop-react-ui/shared', () => ({
  formatCurrency: vi.fn(price => `$${price.toFixed(2)}`),
}));

describe('ProductDetail', () => {
  const mockProduct: ProductResponse = {
    categories: ['category'],
    thumbnail: { name: 'name', uri: 'test-image.jpg' },
    id: 1,
    name: 'Test Product',
    price: 99.99,
    description: 'This is a test product description',
    images: [{ name: 'name', uri: 'test-image.jpg' }],
  };

  beforeEach(() => {
    vi.clearAllMocks();

    getProductByIdSpy.mockReturnValue(mockProduct);
  });

  it('should render successfully', () => {
    const { baseElement } = render(<ProductDetail />);
    expect(baseElement).toBeTruthy();
  });

  it('should match snapshot', () => {
    const { baseElement } = render(<ProductDetail />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should display product details when product is available', () => {
    render(<ProductDetail />);

    expect(screen.getByText('Test Product')).toBeTruthy();
    expect(screen.getByText('$99.99')).toBeTruthy();
    expect(screen.getByText('This is a test product description')).toBeTruthy();

    const productImage: HTMLImageElement = screen.getByAltText('Test Product');
    expect(productImage).toBeTruthy();
    expect(productImage.src).toContain('test-image.jpg');
  });

  it('should show loading state when product is not available', () => {
    getProductByIdSpy.mockReturnValue(undefined);

    render(<ProductDetail />);

    expect(screen.getByText('Loading...')).toBeTruthy();
  });

  it('should call fetchProductById with the correct ID', () => {
    render(<ProductDetail />);

    expect(fetchProductByIdSpy).toHaveBeenCalledWith(1);
  });

  it('should disable the "Add to cart" button', () => {
    fetchProductByIdSpy.mockReturnValue(undefined);
    render(<ProductDetail />);

    const addToCartButton: HTMLButtonElement = screen.getByText('Add to cart');
    expect(addToCartButton.disabled).toBe(true);
  });

  it('should display the placeholder image if the product image fails to load', async () => {
    render(<ProductDetail />);

    const productImage: HTMLImageElement = screen.getByAltText('Test Product');
    expect(productImage).toBeTruthy();

    productImage.dispatchEvent(new Event('error'));
    expect(productImage.src).toContain('images/placeholder-image.jpg');
  });
});
