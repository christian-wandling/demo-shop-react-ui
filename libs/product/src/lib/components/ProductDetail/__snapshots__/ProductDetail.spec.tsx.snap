// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProductDetail > should match snapshot 1`] = `
<body>
  <div>
    <div
      class="product-detail pt-6"
    >
      <div
        class="mx-auto mt-6 max-w-2xl sm:px-6 lg:grid lg:max-w-7xl lg:grid-cols-3 lg:gap-x-8 lg:px-8"
      >
        <div
          class="aspect-h-1 aspect-w-1 overflow-hidden sm:rounded-lg lg:block"
        >
          <img
            alt="Test Product"
            class="h-full w-full object-cover object-center group-hover:opacity-75"
            src="test-image.jpg"
          />
        </div>
      </div>
      <div
        class="mx-auto max-w-2xl px-4 pb-16 pt-10 sm:px-6 lg:grid lg:max-w-7xl lg:grid-cols-3 lg:grid-rows-[auto,auto,1fr] lg:gap-x-8 lg:px-8 lg:pb-24 lg:pt-16"
      >
        <div
          class="lg:col-span-2 lg:border-r lg:border-gray-200 lg:pr-8"
        >
          <h1
            class="text-2xl font-bold tracking-tight text-gray-900 sm:text-3xl"
          >
            Test Product
          </h1>
        </div>
        <div
          class="mt-4 lg:row-span-3 lg:mt-0"
        >
          <h2
            class="sr-only"
          >
            Product information
          </h2>
          <p
            class="text-3xl tracking-tight text-gray-900"
          >
            $99.99
          </p>
          <button
            class="btn-add mt-10 flex w-full items-center justify-center rounded-md border border-transparent px-8 py-3 text-base font-medium
  bg-gray-100 text-gray-800 pointer-events-none"
            disabled=""
          >
            Add to cart
          </button>
        </div>
        <div
          class="py-10 lg:col-span-2 lg:col-start-1 lg:border-r lg:border-gray-200 lg:pb-16 lg:pr-8 lg:pt-6"
        >
          <div
            class="product-description text-sm text-gray-600"
          >
            This is a test product description
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
